### PR TITLE
refactor: 💡 adjust css selector for specific asserts

### PR DESCRIPTION
**Pull request by lisbet-alvarez**
✅ Closes: ICU-6889

:tickets: [Jira ticket](https://hashicorp.atlassian.net/browse/ICU-6889)

## Description
Refactor css selector to be more specific for assert statements that use .doesNotExist() method

### CODE REVIEW

File: ui/admin/tests/acceptance/credential-store/credentials/list-test.js [hunk 1]
```diff
@@ -90,7 +90,7 @@ module('Acceptance | credential-stores | credentials | list', function (hooks) {
         'create'
       )
     );
-    assert.dom(`[href="${urls.credentials}"]`).doesNotExist();
+    assert.dom(`.rose-nav-tabs a[href="${urls.credentials}"]`).doesNotExist();
```
Comment (randallmorey) [inline ui/admin/tests/acceptance/credential-store/credentials/list-test.js:93]:
```suggestion
    assert.dom(`.rose-nav-tabs [href="${urls.credentials}"]`).doesNotExist();
```

File: ui/admin/tests/acceptance/credential-store/credentials/read-test.js [hunk 2]
```diff
@@ -93,7 +93,9 @@ module('Acceptance | credential-stores | credentials | read', function (hooks) {
         (item) => item != 'read'
       );
     await visit(urls.credentials);
-    assert.dom(`[href="${urls.usernamePasswordCredential}"]`).doesNotExist();
+    assert
+      .dom(`.rose-table a[href="${urls.usernamePasswordCredential}"]`)
```
Comment (randallmorey) [inline ui/admin/tests/acceptance/credential-store/credentials/read-test.js:97]:
```suggestion
      .dom(`.rose-table [href="${urls.usernamePasswordCredential}"]`)
```

File: ui/admin/tests/acceptance/credential-store/read-test.js [hunk 3]
```diff
@@ -93,7 +93,9 @@ module('Acceptance | credential-stores | read', function (hooks) {
 
     await click(`[href="${urls.credentialStores}"]`);
 
-    assert.dom(`[href="${urls.staticCredentialStore}"]`).doesNotExist();
+    assert
+      .dom(`.rose-table a[href="${urls.staticCredentialStore}"]`)
```
Comment (randallmorey) [inline ui/admin/tests/acceptance/credential-store/read-test.js:97]:
```suggestion
      .dom(`.rose-table [href="${urls.staticCredentialStore}"]`)
```

File: ui/admin/tests/acceptance/credential-store/read-test.js [hunk 4]
```diff
@@ -107,14 +109,18 @@ module('Acceptance | credential-stores | read', function (hooks) {
...
+      .doesNotExist();
     assert.dom(`[href="${urls.staticCredentialStore}"]`).exists();
   });
 
   test('visiting an unknown credential store displays 404 message', async function (assert) {
     assert.expect(2);
     await visit(urls.credentialStores);
-    assert.dom(`[href="${urls.unknownCredentialStore}"]`).doesNotExist();
+    assert
+      .dom(`.rose-table a[href="${urls.unknownCredentialStore}"]`)
```
Comment (randallmorey) [inline ui/admin/tests/acceptance/credential-store/read-test.js:122]:
```suggestion
      .dom(`.rose-table [href="${urls.unknownCredentialStore}"]`)
```

File: ui/admin/tests/acceptance/host-catalogs/host-sets/read-test.js [hunk 5]
```diff
@@ -86,13 +86,13 @@ module('Acceptance | host-catalogs | host sets | read', function (hooks) {
 
     await click(`[href="${urls.hostSets}"]`);
 
-    assert.dom(`[href="${urls.hostSet}"]`).doesNotExist();
+    assert.dom(`.rose-table a[href="${urls.hostSet}"]`).doesNotExist();
```
Comment (randallmorey) [inline ui/admin/tests/acceptance/host-catalogs/host-sets/read-test.js:89]:
```suggestion
    assert.dom(`.rose-table [href="${urls.hostSet}"]`).doesNotExist();
```

File: ui/admin/tests/acceptance/host-catalogs/host-sets/read-test.js [hunk 6]
```diff
@@ -86,13 +86,13 @@ module('Acceptance | host-catalogs | host sets | read', function (hooks) {
...
   });
 
   test('visiting an unknown host set displays 404 message', async function (assert) {
     assert.expect(2);
     await visit(urls.hostSets);
-    assert.dom(`[href="${urls.unknownHostSet}"]`).doesNotExist();
+    assert.dom(`.rose-table a[href="${urls.unknownHostSet}"]`).doesNotExist();
```
Comment (randallmorey) [inline ui/admin/tests/acceptance/host-catalogs/host-sets/read-test.js:95]:
```suggestion
    assert.dom(`.rose-table [href="${urls.unknownHostSet}"]`).doesNotExist();
```

File: ui/admin/tests/acceptance/host-catalogs/hosts/read-test.js [hunk 7]
```diff
@@ -86,13 +86,13 @@ module('Acceptance | host-catalogs | hosts | read', function (hooks) {
 
     await click(`[href="${urls.hosts}"]`);
 
-    assert.dom(`[href="${urls.host}"]`).doesNotExist();
+    assert.dom(`.rose-table a[href="${urls.host}"]`).doesNotExist();
```
Comment (randallmorey) [inline ui/admin/tests/acceptance/host-catalogs/hosts/read-test.js:89]:
```suggestion
    assert.dom(`.rose-table [href="${urls.host}"]`).doesNotExist();
```

File: ui/admin/tests/acceptance/host-catalogs/hosts/read-test.js [hunk 8]
```diff
@@ -86,13 +86,13 @@ module('Acceptance | host-catalogs | hosts | read', function (hooks) {
...
   });
 
   test('visiting an unknown host displays 404 message', async function (assert) {
     assert.expect(2);
     await visit(urls.hosts);
-    assert.dom(`[href="${urls.unknownHost}"]`).doesNotExist();
+    assert.dom(`.rose-table a[href="${urls.unknownHost}"]`).doesNotExist();
```
Comment (randallmorey) [inline ui/admin/tests/acceptance/host-catalogs/hosts/read-test.js:95]:
```suggestion
    assert.dom(`.rose-table [href="${urls.unknownHost}"]`).doesNotExist();
```

File: ui/admin/tests/acceptance/host-catalogs/read-test.js [hunk 9]
```diff
@@ -78,13 +78,15 @@ module('Acceptance | host-catalogs | read', function (hooks) {
 
     await click(`[href="${urls.hostCatalogs}"]`);
 
-    assert.dom(`[href="${urls.hostCatalog}"]`).doesNotExist();
+    assert.dom(`.rose-table a[href="${urls.hostCatalog}"]`).doesNotExist();
```
Comment (randallmorey) [inline ui/admin/tests/acceptance/host-catalogs/read-test.js:81]:
```suggestion
    assert.dom(`.rose-table [href="${urls.hostCatalog}"]`).doesNotExist();
```

File: ui/admin/tests/acceptance/host-catalogs/read-test.js [hunk 10]
```diff
@@ -78,13 +78,15 @@ module('Acceptance | host-catalogs | read', function (hooks) {
...
   });
 
   test('visiting an unknown host catalog displays 404 message', async function (assert) {
     assert.expect(2);
     await visit(urls.hostCatalogs);
-    assert.dom(`[href="${urls.unknownHostCatalog}"]`).doesNotExist();
+    assert
+      .dom(`.rose-table a[href="${urls.unknownHostCatalog}"]`)
```
Comment (randallmorey) [inline ui/admin/tests/acceptance/host-catalogs/read-test.js:88]:
```suggestion
      .dom(`.rose-table [href="${urls.unknownHostCatalog}"]`)
```

File: ui/admin/tests/acceptance/sessions-test.js [hunk 11]
```diff
@@ -94,7 +94,8 @@ module('Acceptance | sessions', function (hooks) {
         'list'
       )
     );
-    assert.dom(`[href="${urls.sessions}"]`).doesNotExist();
+
+    assert.dom(`nav:nth-child(2) a[href="${urls.sessions}"]`).doesNotExist();
```
Comment (randallmorey) [inline ui/admin/tests/acceptance/sessions-test.js:98]:
```suggestion
    assert.dom(`nav:nth-child(2) [href="${urls.sessions}"]`).doesNotExist();
```

File: ui/admin/tests/acceptance/targets/brokered-credential-sources-test.js [hunk 12]
```diff
@@ -236,7 +236,11 @@ module('Acceptance | targets | brokered credential sources', function (hooks) {
         (item) => item !== 'add-credential-sources'
       );
     await visit(urls.brokeredCredentialSources);
-    assert.dom(`[href="${urls.addBrokeredCredentialSources}"]`).doesNotExist();
+    assert
+      .dom(
+        `.rose-dropdown-content a[href="${urls.addBrokeredCredentialSources}"]`
```
Comment (randallmorey) [inline ui/admin/tests/acceptance/targets/brokered-credential-sources-test.js:241]:
```suggestion
        `.rose-dropdown-content [href="${urls.addBrokeredCredentialSources}"]`
```

File: ui/admin/tests/acceptance/targets/read-test.js [hunk 13]
```diff
@@ -94,7 +94,7 @@ module('Acceptance | targets | read', function (hooks) {
 
     await click(`[href="${urls.targets}"]`);
 
-    assert.dom(`[href="${urls.sshTarget}"]`).doesNotExist();
+    assert.dom(`.rose-table a[href="${urls.sshTarget}"]`).doesNotExist();
```
Comment (randallmorey) [inline ui/admin/tests/acceptance/targets/read-test.js:97]:
```suggestion
    assert.dom(`.rose-table [href="${urls.sshTarget}"]`).doesNotExist();
```

File: ui/admin/tests/acceptance/targets/read-test.js [hunk 14]
```diff
@@ -106,14 +106,14 @@ module('Acceptance | targets | read', function (hooks) {
 
     await click(`[href="${urls.targets}"]`);
 
-    assert.dom(`[href="${urls.tcpTarget}"]`).doesNotExist();
+    assert.dom(`.rose-table a[href="${urls.tcpTarget}"]`).doesNotExist();
```
Comment (randallmorey) [inline ui/admin/tests/acceptance/targets/read-test.js:109]:
```suggestion
    assert.dom(`.rose-table [href="${urls.tcpTarget}"]`).doesNotExist();
```

File: ui/admin/tests/acceptance/targets/read-test.js [hunk 15]
```diff
@@ -106,14 +106,14 @@ module('Acceptance | targets | read', function (hooks) {
...
     assert.dom(`[href="${urls.sshTarget}"]`).exists();
   });
 
   test('visiting an unknown target displays 404 message', async function (assert) {
     assert.expect(2);
     await visit(urls.targets);
-    assert.dom(`[href="${urls.unknownTarget}"]`).doesNotExist();
+    assert.dom(`.rose-table a[href="${urls.unknownTarget}"]`).doesNotExist();
```
Comment (randallmorey) [inline ui/admin/tests/acceptance/targets/read-test.js:116]:
```suggestion
    assert.dom(`.rose-table [href="${urls.unknownTarget}"]`).doesNotExist();
```

**Review by randallmorey — CHANGES_REQUESTED** [review]

File: ui/admin/tests/acceptance/credential-store/credentials/read-test.js [hunk 16]
```diff
@@ -103,7 +105,9 @@ module('Acceptance | credential-stores | credentials | read', function (hooks) {
         (item) => item != 'read'
       );
     await visit(urls.credentials);
-    assert.dom(`[href="${urls.usernameKeyPairCredential}"]`).doesNotExist();
+    assert
+      .dom(`.rose-table a[href="${urls.usernameKeyPairCredential}"]`)
```
Comment (cameronperera) [inline ui/admin/tests/acceptance/credential-store/credentials/read-test.js:109]:
```suggestion
      .dom(`.rose-table [href="${urls.usernameKeyPairCredential}"]`)
```

File: ui/admin/tests/acceptance/credential-store/read-test.js [hunk 17]
```diff
@@ -107,14 +109,18 @@ module('Acceptance | credential-stores | read', function (hooks) {
 
     await click(`[href="${urls.credentialStores}"]`);
 
-    assert.dom(`[href="${urls.vaultCredentialStore}"]`).doesNotExist();
+    assert
+      .dom(`.rose-table a[href="${urls.vaultCredentialStore}"]`)
```
Comment (cameronperera) [inline ui/admin/tests/acceptance/credential-store/read-test.js:113]:
```suggestion
      .dom(`.rose-table [href="${urls.vaultCredentialStore}"]`)
```

**Review by calcaide — CHANGES_REQUESTED** [review]
Everything LGTM but I do agree on Cameron's suggestiong to change.

**Comment by ZedLi** [comment]
Sorry what was the issue with checking for the attribute on the `a` element?

Unrelated to that issue, I actually don't think there's too much value in making it more specific just to the table. I think it's the value is in checking for the specific row we're expecting it to be in, though I can see how that much might be much harder to change for all of these.

**Comment by randallmorey** [comment]
> Sorry what was the issue with checking for the attribute on the `a` element?
> 
> Unrelated to that, I actually don't think there's too much value in making it more specific just to the table either way. I think the issue still remains in that if we mess up the URL/href, we might be _thinking_ we're checking that it doesn't exist correctly but really we just got the URL wrong.
> 
> I think what is a more accurate test would be checking specifically for an `a` element (with no url attribute) inside a row/cell of the table. There shouldn't be an `a` link if we don't have permission.

Ah good point.  Proving absence needs more thought.

For selectors, it's generally better to use the least specific selector possible.  In CSS the reason is to avoid a specificity race.  In JavaScript, less specific selectors are less likely to break when markup changes are made.

Ultimately we need to rethink test selectors.  I know a number of other product teams use `data-` attributes to directly mark elements of interest.

**Comment by ZedLi** [comment]
> For selectors, it's generally better to use the least specific selector possible. In CSS the reason is to avoid a specificity race. In JavaScript, less specific selectors are less likely to break when the markup changes.
> 
> Ultimately we need to rethink test selectors. I know a number of other product teams use `data-` attributes to directly mark elements of interest.

Yup totally agree with both points here, I think that's why I was originally thinking of just checking for `a`. I think this is a good topic to have in tech time, I added it as I do think we should talk about it more

**Comment by lisbet-alvarez** [comment]
> > For selectors, it's generally better to use the least specific selector possible. In CSS the reason is to avoid a specificity race. In JavaScript, less specific selectors are less likely to break when the markup changes.
> > Ultimately we need to rethink test selectors. I know a number of other product teams use `data-` attributes to directly mark elements of interest.
> 
> Yup totally agree with both points here, I think that's why I was originally thinking of just checking for `a` rather than a URL. I think this is a good topic to have in tech time, I added it as I do think we should talk about it more

Very good points were brought up. For now, I will convert this to a draft PR and I will attend tomorrow's tech time to discuss this topic.

**Review by calcaide — APPROVED** [review]
LGTM!

As mentioned in slack, this PR covers what is need and we will further discussion about test selectors.

File: addons/core/tests/integration/helpers/relative-datetime-live-test.js [hunk 18]
```diff
@@ -39,7 +39,7 @@ module('Integration | Helper | relative-datetime-live', function (hooks) {
     await render(hbs`{{relative-datetime-live this.date}}`);
     assert.dom(this.element).hasText('in 7 months');
 
-    this.date = now + 23 * 12 * 31 * 24 * 60 * 60 * 1000;
+    this.date = now + 23 * 12 * 30 * 24 * 60 * 60 * 1000;
```
Comment (lisbet-alvarez) [inline addons/core/tests/integration/helpers/relative-datetime-live-test.js:42]:
```suggestion
    this.date = now + 23 * 12 * 31 * 24 * 60 * 60 * 1000;
```

**Review by ZedLi — APPROVED** [review]
LGTM, thanks for this refactor!

**Review by cameronperera — APPROVED** [review]

**Pull request merged**